sky/utils/controller_utils.py

```diff
@@ -139,34 +139,54 @@ def from_name(cls, name: Optional[str]) -> Optional['Controllers']:
-    commands = [
-        # aws
-        'pip list | grep boto3 > /dev/null 2>&1 || '
-        'pip install "urllib3<2" awscli>=1.27.10 botocore>=1.29.10 '
-        'boto3>=1.26.1 > /dev/null 2>&1',
-        # gcp
-        'pip list | grep google-api-python-client > /dev/null 2>&1 || '
-        'pip install google-api-python-client>=2.69.0 google-cloud-storage '
-        '> /dev/null 2>&1',
-        f'{gcp.GOOGLE_SDK_INSTALLATION_COMMAND}',
-        # fluidstack does not need to install any cloud dependencies.
-    ]
-    # k8s and ibm doesn't support open port and spot instance yet, so we don't
-    # install them for either controller.
-    if controller_type == 'spot':
-        # oci doesn't support open port yet, so we don't install oci
-        # dependencies for sky serve controller.
-        commands.append('pip list | grep oci > /dev/null 2>&1 || '
-                        'pip install oci > /dev/null 2>&1')
+    commands = []
+    enabled_clouds = sky_check.get_cached_enabled_clouds_or_refresh()
     # TODO(tian): Make dependency installation command a method of cloud
     # class and get all installation command for enabled clouds.
-    if any(
-            cloud.is_same_cloud(clouds.Azure())
-            for cloud in sky_check.get_cached_enabled_clouds_or_refresh()):
+    # AWS
+    if clouds.cloud_in_list(clouds.AWS(), enabled_clouds):
+        commands.append(
+            'pip list | grep boto3 > /dev/null 2>&1 || '
+            'pip install "urllib3<2" awscli>=1.27.10 botocore>=1.29.10 '
+            'boto3>=1.26.1 > /dev/null 2>&1')
+    # GCP
+    if clouds.cloud_in_list(clouds.GCP(), enabled_clouds):
+        commands.extend([
+            'pip list | grep google-api-python-client > /dev/null 2>&1 || '
+            'pip install google-api-python-client>=2.69.0 google-cloud-storage '
+            '> /dev/null 2>&1', f'{gcp.GOOGLE_SDK_INSTALLATION_COMMAND}'
+        ])
+    # Azure
+    if clouds.cloud_in_list(clouds.Azure(), enabled_clouds):
         commands.append(
             'pip list | grep azure-cli > /dev/null 2>&1 || '
             'pip install azure-cli>=2.31.0 azure-core azure-identity>=1.13.0 '
             'azure-mgmt-network > /dev/null 2>&1')
+    # Kubernetes
+    if clouds.cloud_in_list(clouds.Kubernetes(), enabled_clouds):
+        commands.append(
+            # Install k8s + skypilot dependencies
+            'sudo bash -c "if '
+            '! command -v curl &> /dev/null || '
+            '! command -v socat &> /dev/null || '
+            '! command -v netcat &> /dev/null; '
+            'then apt update && apt install curl socat netcat -y; '
+            'fi" && '
+            # Install kubectl
+            '(command -v kubectl &>/dev/null || '
+            '(curl -LO "https://dl.k8s.io/release/$(curl -L -s '
+            'https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl" && '
+            'sudo install -o root -g root -m 0755 '
+            'kubectl /usr/local/bin/kubectl)) && ')
```

Comment:
Should we consider adding these dependencies in our default kubernetes image?

Reply:
Tracked in https://github.com/skypilot-org/skypilot/issues/3435